internal/services/discovery/discovery.go

```diff
@@ -0,0 +1,168 @@
+
+	// GetKOPSClusterNameAndStateStore discovers the cluster name and kOps state store bucket from the kube-system namespace
+	// annotation. kOps annotates the kube-system namespace with annotations such as this:
+	// * addons.k8s.io/core.addons.k8s.io: '{"version":"1.4.0","channel":"s3://bucket/cluster-name/addons/bootstrap-channel.yaml","manifestHash":"hash"}'
+	// We can retrieve the state store bucket name and the cluster name from the "channel" property of the annotation value.
+	GetKOPSClusterNameAndStateStore(ctx context.Context, log logrus.FieldLogger) (clusterName, stateStore *string, reterr error)
+
+	// GetOpenshiftClusterID discovers the OpenShift cluster ID by reading the cluster ID from the OpenShift cluster version resource.
+	GetOpenshiftClusterID(ctx context.Context) (*string, error)
+
+	// GetOpenshiftClusterName discovers the OpenShift cluster name by reading the cluster name from the OpenShift master machine resource.
+	GetOpenshiftClusterName(ctx context.Context) (*string, error)
+}
+
+var _ Service = (*ServiceImpl)(nil)
+
+type ServiceImpl struct {
+	clientset kubernetes.Interface
+	dyno      dynamic.Interface
+
+	kubeSystemNamespace   *v1.Namespace
+	kubeSystemNamespaceMu *sync.Mutex
+}
+
+func New(clientset kubernetes.Interface, dyno dynamic.Interface) *ServiceImpl {
+	return &ServiceImpl{
+		clientset:             clientset,
+		dyno:                  dyno,
+		kubeSystemNamespaceMu: &sync.Mutex{},
+	}
+}
+
+func (s *ServiceImpl) GetCSPAndRegion(ctx context.Context) (csp, region *string, reterr error) {
+	return s.getCSPAndRegion(ctx, "")
+}
+
+func (s *ServiceImpl) GetClusterID(ctx context.Context) (*uuid.UUID, error) {
+	ns, err := s.getKubeSystemNamespace(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterID, err := uuid.Parse(string(ns.UID))
+	if err != nil {
+		return nil, fmt.Errorf("parsing namespace %q uid: %w", metav1.NamespaceSystem, err)
+	}
+
+	return &clusterID, nil
+}
+
+func (s *ServiceImpl) getKubeSystemNamespace(ctx context.Context) (*v1.Namespace, error) {
+	s.kubeSystemNamespaceMu.Lock()
+	defer s.kubeSystemNamespaceMu.Unlock()
+
+	if s.kubeSystemNamespace != nil {
+		return s.kubeSystemNamespace, nil
+	}
+
+	ns, err := s.clientset.CoreV1().Namespaces().Get(ctx, metav1.NamespaceSystem, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("getting namespace %q: %w", metav1.NamespaceSystem, err)
+	}
+
+	s.kubeSystemNamespace = ns
+
+	return ns, nil
+}
+
+func (s *ServiceImpl) getCSPAndRegion(ctx context.Context, next string) (csp, region *string, reterr error) {
+	nodes, err := s.clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{Limit: 10, Continue: next})
+	if err != nil {
+		return nil, nil, fmt.Errorf("listing nodes: %w", err)
+	}
+
+	for i, n := range nodes.Items {
+		ready := false
+
+		for _, cond := range n.Status.Conditions {
+			if cond.Type == v1.NodeReady && cond.Status == v1.ConditionTrue {
+				ready = true
+				break
+			}
+		}
+
+		if !ready {
+			continue
+		}
```

Comment:
could be some isNodeReady helper

Reply:
Also, more of a question, why can we only check the region/csp on ready nodes?

Reply:
I guess as agent is running, there will be at least one ready node, so maybe they're safer to use:)

Reply:
seems like it's a decision from earlier code, only moved here

Reply:
Node objects can be created freely but status conditions are managed by kubelet. We probably got cought on failing to find the data due to some zombie object in the past. Left the readiness check but moved it to a func.